server/jetstream_cluster.go

```diff
@@ -4172,8 +4434,79 @@ func (s *Server) jsClusteredStreamUpdateRequest(ci *ClientInfo, acc *Account, su
 			}
 			rg.Peers = peers
 		} else {
-			// We are deleting nodes here.
-			rg.Peers = rg.Peers[:newCfg.Replicas]
+			// We are deleting nodes here. We want to do our best to preserve the current leader.
+			// We have support now from above that guarantees we are in our own Go routine, so can
+			// ask for stream info from the stream leader to make sure we keep the leader in the new list.
+			var curLeader string
+			if !s.allPeersOffline(rg) {
+				// Need to release js lock.
+				js.mu.Unlock()
+				s.mu.Lock()
+				inbox := s.newRespInbox()
+				results := make(chan *StreamInfo, 1)
+				// Store our handler.
+				s.sys.replies[inbox] = func(sub *subscription, _ *client, _ *Account, subject, _ string, msg []byte) {
+					var si StreamInfo
+					if err := json.Unmarshal(msg, &si); err != nil {
+						s.Warnf("Error unmarshaling clustered stream info response:%v", err)
+						return
+					}
+					select {
+					case results <- &si:
+					default:
+						s.Warnf("Failed placing remote stream info result on internal channel")
+					}
+					// Only expect one so clean up here.
+					s.mu.Lock()
```

Comment:
What if there is no response received? We are going to "leak" the inbox in s.sys.replies, no? That is ok if that does not happen often, but I have no idea if that is the case. So should this code be moved out and done at the end of this function (either because the reply was received or we timeout)?

Reply:
Agree. Fixed.